### DISCUSSION TITLE
question about uv__io_stop and uv__io_close

### DISCUSSION THREAD

**Discussion by theanarkh** [libuv/libuv#3801] [Q&A] [answered]
Hi, i have a question. why `uv__io_close` removes the `fd` from `epoll` synchronously, but `uv__io_stop` just set `loop->watchers[w->fd] = NULL` and removes  the`fd` after `epoll_wait`, the related code is as follows.
```
w = loop->watchers[fd];
if (w == NULL) {
  epoll_ctl(loop->backend_fd, EPOLL_CTL_DEL, fd, pe);
  continue;
}
```
1. if we call `uv__io_start` with `POLLIN` at first.
2. then we call `uv__io_stop` with `POLLIN` after `epoll_ctl` or `epoll_wait`.
3. when the `POLLIN` event of `fd` is triggered, the process will still wake up and remove the `fd` because `loop->watchers[w->fd]` is `NULL`.

is it right ? Thanks for any help.

**Comment by bnoordhuis** [comment, marked as answer]
Premise:

1. `uv__io_close(fd)` means libuv is going to close the file descriptor, whereas

2. `uv__io_stop(fd)` means libuv expects to call `uv__io_start(fd)` again in the future

The reason to eagerly deregister the file descriptor in (1) is because of an epoll subtlety.

A file descriptor that's closed in _this_ process but alive in _another_ process still generates epoll events in _this_ process - but the file descriptor is closed so it's no longer possible to remove it from the epoll set.

That isn't a big deal with edge-triggered i/o - libuv sees a spurious event it could simply ignore - but it leads to spinning (busy looping) with level-triggered i/o because the kernel keeps reporting the event over and over again.

It's not a problem with (2) because the file descriptor is still alive. Libuv can deregister it lazily. Saves a system call in the common case.

**Comment by theanarkh** [comment]
Thank you very much.
1 
> A file descriptor that's closed in this process but alive in another process still generates epoll events in this process.

I'm surprised to know that.

2 
After reading your reply, I understand that：

uv__io_stop: `uv__io_stop` mean `fd` is alive. So if we don not call `uv__io_start` after `uv__io_stop`, `Libuv` will remove it when `epoll_wait` return this fd, otherwise `epoll_wait` will return this fd over and over again because Libuv is `level-triggered`.

uv__io_close: after `uv__io_close`, `Libuv` will close the `fd`, but `epoll_wait` maybe return this `fd` over and over again  If the `fd` is alive in other process, so we need to remove it in `uv__io_close`.

3
> but the file descriptor is closed so it's no longer possible to remove it from the epoll set.
Is this code why we can not remove the closed `fd` ?
```
int do_epoll_ctl(int epfd, int op, int fd, struct epoll_event *epds, bool nonblock)
{
        int error;
        struct fd f;
	error = -EBADF;
	f = fdget(epfd);
	if (!f.file)
		goto error_return;
        // ...
}
```

**Reply by bnoordhuis** [reply]
Yes, that's an accurate summary and that's the relevant kernel code.
